### DISCUSSION TITLE
交叉编译BCM4709C构架下的alist出现go-sqlite3 requires cgo to work错误解决

### DISCUSSION THREAD

**Discussion by Hermit-Nameless** [AlistGo/alist#2237] [Q&A] [answered]
@Xhofe 我开始为斐讯K3（arm BCM4709C，梅林固件）交叉编译alist
编译环境：Windows 11 Linux子系统
环境变量：GOOS=linux GOARCH=arm GOARM=5 CGO_ENABLED=0
这样编译出来的程序可以在路由上运行，但运行到init logrus后就退出，查看log.log文件，有failed to connect database:Binary was compiled with 'CGO_ENABLED=0', go-sqlite3 requires cgo to work. This is a stub提示。
我将CGO_ENABLED改为1，并配置CC和CXX值为arm-linux-musleabi-cross或arm-linux-musleabihf-cross以及armv?的其它各个版本的现成库（[https://musl.cc/](https://musl.cc/)下载的）多次编译，得到的可执行文件又都不能在K3梅林上运行了，都是提示not found，请问我的编译环境还需要加入什么库文件吗，还是我配置的CC和CXX库不对？谢谢~
![QQ截图20221107095750](https://user-images.githubusercontent.com/30073216/200213087-518987a6-59b4-4457-9cfb-9c6944e2f5cc.jpg)

**Comment by Hermit-Nameless** [comment, marked as answer]
我为斐讯K3交叉编译成功了，希望这些笔记对想自己编译的同学有所帮助：
Alist要用到sqlite3，需为go安装go-sqlite3驱动
#以下指令要在Alist项目目录下才能正确运行
安装go get：go get
安装go-sqlite3：go get github.com/mattn/go-sqlite3@latest       //需要配合下面CGO_ENABLED、CGO_LDFLAGS、CC、CXX、AR配置的环境变量将sqlite3静态编译到Alist才能正常工作，各平台musl GCC编译器到[https://musl.cc/](https://musl.cc/)下载

为go配置环境变量（前三条跨平台交叉编译才需要配置，编译本地程序默认即可）：
export GOOS=linux                                          //指定交叉编译生成的可执行文件之目标系统
export GOARCH=arm                                          //指定交叉编译生成的可执行文件之目标CPU构架
export GOARM=5                                             //指定交叉编译生成的可执行文件之目标CPU类型，仅在GOARCH=arm时有效
export CGO_ENABLED=1                                       //开启使用GCC编译器
export CGO_LDFLAGS="-static"                               //一定要将目标平台对应的musl GCC库（如斐讯K3就用arm-linux-musleabi-cross）静态编译到可执行文件中，否则到目标平台运行大概率会出现not found错误
export CC=/...保存路径.../arm-linux-musleabi-cross/bin/arm-linux-musleabi-gcc           //为go-sqlite3指定目标平台对应的musl GCC编译器，下载的包解压出来，要指定编译器文件的路径
export CXX=/...保存路径.../arm-linux-musleabi-cross/bin/arm-linux-musleabi-g++
export AR=/...保存路径.../arm-linux-musleabi-cross/bin/arm-linux-musleabi-ar

然后根据Alist文档的编译方法编译：
https://alist.nn.ci/zh/guide/install/source.html

**Reply by anwen-anyi** [reply]
emmm 这个可以在Windows交叉编译Linux二进制的吗？ 我之前试过编译好像也是报   **go-sqlite3** 相关的问题...

**Reply by Hermit-Nameless** [reply]
简单看了一些说明，musl好像不支持在Windows下编译，你得多找找相关的技术文档。
不过现在Windows 10以上的系统，Alist-Web我直接用作者编译好的Releases版，只要再用WSL构建一个Linux下的golang+gcc环境编译后端也很简单。

**Reply by gsoyond** [reply]
嗯，感谢提供的方法，根据步骤编译的alist在netgear R7000上面成功运行了。我之前的经历和你一模一样，下载各种版本的cross工具链，编译出来的都没运行成功，通过log能找到是sqlite3的问题，但一时没想到什么解决办法。你的方法对我有有效，再次感谢！

**Reply by evan618** [reply]
请问一下你选择的是哪个版本的编译工具，我尝试编译后运行都报 “ Illegal instruction” 应该是指令集或者工具没选对。

**Reply by Hermit-Nameless** [reply]
go version是1.19.3，大概率是那几个export环境变量配置错了，不能照抄我的呀，要根据你实际的运行环境定义交叉编译的参数。
